ecdsa/signing/finalize.go

```diff
@@ -7,73 +7,323 @@
+		return nil, nil, FinalizeWrapError(errors.New("identify abort assertion fail in phase 7"), ourP, culprits...)
 	}
 
-	recid := 0
+	// Calculate Recovery ID: It is not possible to compute the public key out of the signature itself;
+	// the Recovery ID is used to enable extracting the public key from the signature.
 	// byte v = if(R.X > curve.N) then 2 else 0) | (if R.Y.IsEven then 0 else 1);
-	if round.temp.rx.Cmp(tss.EC().Params().N) > 0 {
-		recid = 2
+	recId := 0
+	if bigR.X().Cmp(N) > 0 {
+		recId = 2
 	}
-	if round.temp.ry.Bit(0) != 0 {
-		recid |= 1
+	if bigR.Y().Bit(0) != 0 {
+		recId |= 1
 	}
 
 	// This is copied from:
 	// https://github.com/btcsuite/btcd/blob/c26ffa870fd817666a857af1bf6498fabba1ffe3/btcec/signature.go#L442-L444
 	// This is needed because of tendermint checks here:
 	// https://github.com/tendermint/tendermint/blob/d9481e3648450cb99e15c6a070c1fb69aa0c255b/crypto/secp256k1/secp256k1_nocgo.go#L43-L47
-	secp256k1halfN := new(big.Int).Rsh(tss.EC().Params().N, 1)
-	if sumS.Cmp(secp256k1halfN) > 0 {
-		sumS.Sub(tss.EC().Params().N, sumS)
-		recid ^= 1
+	secp256k1halfN := new(big.Int).Rsh(N, 1)
+	if s.Cmp(secp256k1halfN) > 0 {
+		s.Sub(N, s)
+		recId ^= 1
+	}
+
+	ok := ecdsa.Verify(pk, msg.Bytes(), r, s)
+	if !ok {
+		return nil, nil, FinalizeWrapError(fmt.Errorf("signature verification 1 failed"), ourP)
 	}
 
 	// save the signature for final output
-	round.data.Signature = append(round.temp.rx.Bytes(), sumS.Bytes()...)
-	round.data.SignatureRecovery = []byte{byte(recid)}
-	round.data.R = round.temp.rx.Bytes()
-	round.data.S = sumS.Bytes()
-	round.data.M = round.temp.m.Bytes()
+	signature := new(common.ECSignature)
+	signature.R, signature.S = r.Bytes(), s.Bytes()
+	signature.Signature = append(r.Bytes(), s.Bytes()...)
+	signature.SignatureRecovery = []byte{byte(recId)}
+	signature.M = msg.Bytes()
+	state.Signature = signature
+
+	btcecSig := &btcec.Signature{R: r, S: s}
+	if ok = btcecSig.Verify(msg.Bytes(), (*btcec.PublicKey)(pk)); !ok {
+		return nil, nil, FinalizeWrapError(fmt.Errorf("signature verification 2 failed"), ourP)
+	}
+
+	// SECURITY: to be safe the oneRoundData is no longer needed here and reuse of `r` can compromise the key
+	state.OneRoundData = nil
+
+	return state, btcecSig, nil
+}
+
+func FinalizeWrapError(err error, victim *tss.PartyID, culprits ...*tss.PartyID) *tss.Error {
+	return tss.NewError(err, TaskNameFinalize, 8, victim, culprits...)
+}
 
-	pk := ecdsa.PublicKey{
+// -----
+// Full Online Finalization &
+// Identify Aborts of "Type 7"
+// ------
+
+func (round *finalization) Start() *tss.Error {
+	if round.started {
+		return round.WrapError(errors.New("round already started"))
+	}
+	round.number = 8
+	round.started = true
+	round.resetOK()
+
+	Ps := round.Parties().IDs()
+	Pi := round.PartyID()
+	i := Pi.Index
+
+	culprits := make([]*tss.PartyID, 0, len(round.temp.signRound6Messages))
+
+	// Identifiable Abort Type 7 triggered during Phase 6 (GG20)
+	if round.abortingT7 {
```

Comment:
how about making this check part of online processing? Because both `FinalizeGetAndVerifyFinalSig` and `FinalizeGetOurSigShare` are unable to do the check without context if we separate online and offline signing:

```
r1msg1 := round.temp.signRound1Message1s[j].Content().(*SignRound1Message1)
if err != nil || !bytes.Equal(cA.Bytes(), r1msg1.GetC()) {
	culprits = append(culprits, Pj)
	continue
}
```

Reply:
This check is already and only apply in the preparation, thus it is safe here to get the paillier encryption of the `k_i`. 

This check is for the gg20 signing phrase6, so it should be done online, and only check once abort is set in [https://github.com/binance-chain/tss-lib/blob/aa7164ab690d7b494bce75c163866ddff75fe2cd/ecdsa/signing/round_7.go#L167](url)

if you are doing one-round-signing, this check will not been run(already checked in preparation).